concept/thing/impl/AttributeImpl.java

```diff
@@ -41,6 +42,7 @@
 public abstract class AttributeImpl<VALUE> extends ThingImpl implements Attribute {
 
     AttributeVertex<VALUE> attributeVertex;
+    AttributeTypeImpl attributeType;
```

Comment:
we introduce a cache of attribute type to make calls to things like `getValueType()` fast, which would occur many times when sorting